From 0754a8f217855090e6a9aee63d9916eaf739aa71 Mon Sep 17 00:00:00 2001
From: Ricardo Salveti <ricardo@foundries.io>
Date: Mon, 4 Oct 2021 21:23:18 -0300
Subject: [PATCH 257/322] [FIO toup-squash] spl: use spl_simple_fit_parse for
 STRICT validation

FIT image is already parsed and verified by the spl_simple_fit_parse
function, so instead of searching for the default config node entry and
validating its signature, just use the return code from the fit parser
and hang in case of errors.

Fixes: dd9336eccb ("[FIO fromlist] spl: Add CONFIG_SPL_FIT_SIGNATURE_STRICT")
Signed-off-by: Ricardo Salveti <ricardo@foundries.io>
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>
---
 common/spl/spl_fit.c | 24 ++++++++----------------
 1 file changed, 8 insertions(+), 16 deletions(-)

diff --git a/common/spl/spl_fit.c b/common/spl/spl_fit.c
index d22ad97bd17..62dd3d5fee4 100644
--- a/common/spl/spl_fit.c
+++ b/common/spl/spl_fit.c
@@ -635,27 +635,19 @@ int spl_load_simple_fit(struct spl_image_info *spl_image,
 	if (ret < 0)
 		return ret;
 
-	if (CONFIG_IS_ENABLED(FIT_SIGNATURE_STRICT)) {
-		int cfg_noffset = fit_conf_get_node(fit, NULL);
-
-		if (cfg_noffset >= 0) {
-			if (fit_config_verify(fit, cfg_noffset)) {
-				puts("Unable to verify the required FIT config.\n");
-				hang();
-			}
-		} else {
-			puts("SPL_FIT_SIGNATURE_STRICT needs a config node in FIT\n");
-			hang();
-		}
-	}
-
 	/* skip further processing if requested to enable load-only use cases */
 	if (spl_load_simple_fit_skip_processing())
 		return 0;
 
 	ret = spl_simple_fit_parse(&ctx);
-	if (ret < 0)
-		return ret;
+	if (ret < 0) {
+		if (CONFIG_IS_ENABLED(FIT_SIGNATURE_STRICT)) {
+			puts("SPL_FIT_SIGNATURE_STRICT needs a valid config node in FIT\n");
+			hang();
+		} else {
+			return ret;
+		}
+	}
 
 #if defined(CONFIG_DUAL_BOOTLOADER) && defined(CONFIG_IMX_TRUSTY_OS)
     int rbindex;
-- 
2.34.1

